pkgs/development/python-modules/vllm/default.nix

```diff
@@ -34,65 +44,61 @@
   rocmSupport ? config.rocmSupport,
   rocmPackages ? { },
   gpuTargets ? [ ],
-}:
+}@args:
 
 let
-  stdenv_pkg = stdenv;
+  cutlass = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "cutlass";
+    rev = "refs/tags/v3.5.0";
+    sha256 = "sha256-D/s7eYsa5l/mfx73tE4mnFcTQdYqGmXa9d9TCryw4e4=";
+  };
 in
 
 buildPythonPackage rec {
   pname = "vllm";
-  version = "0.3.3";
-  format = "pyproject";
+  version = "0.5.3.post1";
+  pyproject = true;
 
   src = fetchFromGitHub {
     owner = "vllm-project";
     repo = pname;
-    rev = "v${version}";
-    hash = "sha256-LU5pCPVv+Ws9dL8oWL1sJGzwQKI1IFk2A1I6TP9gXL4=";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-++DK2Y2zz+1KrEcdQc5XFrSjc7fCwMD2DQ/RqY7PoFU=";
   };
 
+  stdenv = if cudaSupport then cudaPackages.backendStdenv else args.stdenv;
+
   # Otherwise it tries to enumerate host supported ROCM gfx archs, and that is not possible due to sandboxing.
   PYTORCH_ROCM_ARCH = lib.optionalString rocmSupport (
     lib.strings.concatStringsSep ";" rocmPackages.clr.gpuTargets
   );
 
-  # cupy-cuda12x is the same wheel as cupy, but built with cuda dependencies, we already have it set up
-  # like that in nixpkgs. Version upgrade is due to upstream shenanigans
-  # https://github.com/vllm-project/vllm/pull/2845/commits/34a0ad7f9bb7880c0daa2992d700df3e01e91363
-  #
   # hipcc --version works badly on NixOS due to unresolved paths.
-  # Unclear why pythonRelaxDeps doesn't work here, but on last attempt, it didn't.
   postPatch =
     ''
-      substituteInPlace requirements.txt \
-        --replace "xformers == 0.0.23.post1" "xformers"
-      substituteInPlace requirements.txt \
-        --replace "cupy-cuda12x == 12.1.0" "cupy"
-      substituteInPlace requirements-build.txt \
-        --replace "torch==2.1.2" "torch"
-      substituteInPlace pyproject.toml \
-        --replace "torch == 2.1.2" "torch"
-      substituteInPlace requirements.txt \
-        --replace "torch == 2.1.2" "torch"
+      substituteInPlace setup.py \
+        --replace 'cmake_args = [' 'cmake_args = ["-DFETCHCONTENT_SOURCE_DIR_CUTLASS=${cutlass}",'
     ''
     + lib.optionalString rocmSupport ''
       substituteInPlace setup.py \
         --replace "'hipcc', '--version'" "'${writeShellScript "hipcc-version-stub" "echo HIP version: 0.0"}'"
     '';
 
   preBuild =
-    lib.optionalString cudaSupport ''
-      export CUDA_HOME=${cudaPackages.cuda_nvcc}
-    ''
-    + lib.optionalString rocmSupport ''
+    lib.optionalString rocmSupport ''
       export ROCM_HOME=${rocmPackages.clr}
       export PATH=$PATH:${rocmPackages.hipcc}
+    ''
+    + lib.optionalString cudaSupport ''
+      export CUDA_HOME=${cudaPackages.cuda_nvcc}
     '';
 
   nativeBuildInputs = [
+    cmake
     ninja
     packaging
+    pythonRelaxDepsHook
     setuptools
     torch
```

Comment:
Side note: torch in nativeBuildInputs looks wrong

Reply:
yeah, not sure why that was put there.

Reply:
Maybe one of those 'let the user bring their own `torch`' ideas a la the current `torchmetrics`?

Reply:
> Maybe one of those 'let the user bring their own torch' ideas a la the current torchmetrics?

I rather meant that vllm likely needs the host platform's torch, not the build platform's

Reply:
ok - I think this torch can be removed, torch is already in propagatedBuildInputs, shouldn't be needed twice.